FIAT/functional.py

```diff
@@ -336,22 +339,214 @@ def __init__(self, ref_el, facet_no, Q, f_at_qpts):
+        n = cell.compute_scaled_normal(entity)
+        IntegralLegendreDirectionalMoment.__init__(self, cell, n, entity,
+                                                   mom_deg, comp_deg,
+                                                   "IntegralLegendreNormalMoment")
+
+
+class IntegralLegendreTangentialMoment(IntegralLegendreDirectionalMoment):
+    """Momement of v.t against a Legendre polynomial over an edge"""
+    def __init__(self, cell, entity, mom_deg, comp_deg):
+        t = cell.compute_edge_tangent(entity)
+        IntegralLegendreDirectionalMoment.__init__(self, cell, t, entity,
+                                                   mom_deg, comp_deg,
+                                                   "IntegralLegendreTangentialMoment")
+
+
+class IntegralLegendreBidirectionalMoment(Functional):
+    """Moment of dot(s1, dot(tau, s2)) against Legendre on entity, multiplied by the size of the reference facet"""
+    def __init__(self, cell, s1, s2, entity, mom_deg, comp_deg, nm=""):
+        # mom_deg is degree of moment, comp_deg is the total degree of
+        # polynomial you might need to integrate (or something like that)
+        sd = cell.get_spatial_dimension()
+        shp = (sd, sd)
+
+        s1s2T = numpy.outer(s1, s2)
+        quadpoints = comp_deg + 1
+        Q = GaussLegendreQuadratureLineRule(interval(), quadpoints)
+
+        # The volume squared gets the Jacobian mapping from line interval
+        # and the edge length into the functional.
+        legendre = numpy.polynomial.legendre.legval(2*Q.get_points()-1, [0]*mom_deg + [1]) * numpy.abs(cell.volume_of_subcomplex(1, entity))**2
+
+        f_at_qpts = numpy.array([s1s2T*legendre[i] for i in range(quadpoints)])
+
+        # Map the quadrature points
+        fmap = cell.get_entity_transform(sd-1, entity)
+        mappedqpts = [fmap(pt) for pt in Q.get_points()]
+        mappedQ = QuadratureRule(cell, mappedqpts, Q.get_weights())
+
+        pt_dict = OrderedDict()
+
+        qpts = mappedQ.pts
+        qwts = mappedQ.wts
+
+        for k in range(len(qpts)):
+            pt_cur = tuple(qpts[k])
+            pt_dict[pt_cur] = [(qwts[k] * f_at_qpts[k, i, j], (i, j))
+                               for (i, j) in index_iterator(shp)]
+
+        Functional.__init__(self, cell, shp, pt_dict, {}, nm)
+
+
+class IntegralLegendreNormalNormalMoment(IntegralLegendreBidirectionalMoment):
+    """Moment of dot(n, dot(tau, n)) against Legendre on entity."""
+    def __init__(self, cell, entity, mom_deg, comp_deg):
+        n = cell.compute_normal(entity)
+        IntegralLegendreBidirectionalMoment.__init__(self, cell, n, n,
+                                                     entity, mom_deg, comp_deg,
+                                                     "IntegralNormalNormalLegendreMoment")
+
+
+class IntegralLegendreNormalTangentialMoment(IntegralLegendreBidirectionalMoment):
+    """Moment of dot(n, dot(tau, n)) against Legendre on entity."""
```

Comment:
Same docstring as one above, but different classname.  Is one of these wrong?

Reply:
Copy/paste error.